### ISSUE TITLE
Enhanced integrity check with all dlist and dindex files

### ISSUE THREAD

**Issue by balazer** [duplicati/duplicati#1809]
I propose an enhanced integrity check of the remote backup:

1. List all remote files
2. Download all remote dlist and dindex files
3. Note inconsistencies between dlist and dindex files (note any missing dlist or dindex files)
4. Confirm that all of the remote dblock files that should be present are present, and with the correct sizes

In other words, verify as much of the remote backup's integrity as possible by ensuring that all of the expected files are consistent and present, while not downloading any actual source file blocks.

This would go beyond the existing verify functionality, which can only check a sampled subset of dlist, dindex, and dblock files.

**Comment by renestach** [comment]
Duplicati.CommandLine.exe help verify

Usage: test <storage-URL> [<samples>] [<options>]

  Verifies integrity of a backup. A random sample of dlist, dindex, dblock files is downloaded, decrypted and the content is checked against recorded size values and data hashes. <samples> specifies the number of samples to be tested. **If "all” is specified, all files in the backup will be tested.** This is a rolling check, i.e. when executed another time different samples are verified than in the first run. A sample consists of 1 dlist, 1 dindex, 1 dblock.

  --time=<time>
    Checks samples from a specific time.
  --version=<int>
    Checks samples from specific versions. Delimiters are , -

**Comment by balazer** [comment]
Yes, but the existing verify command would download some or all of the dblock files.  A full integrity check that way would be very slow and generate a lot of network traffic.  Hence I propose an integrity check that would merely check for the presence of the dblock files without downloading all of them.

**Comment by mnaiman** [comment]
Yes, but for that "verify all" you need to have builded local database. Problem with verification is, that it would be beneficial to be able to test remote data without local database. 
One month before my local database was corrupted, so I want to recreate it. But remote storage missed some files. So rebuild was not possible and remote integrity check neither.

**Comment by kenkendk** [comment]
There is a little used feature `--upload-verification-file` which uploads an extra file in json format with the hashes of all files expected to be on the remote backend.

There is a python script that can read this file and verify that all files are present and in the state expected by the database:
https://github.com/duplicati/duplicati/blob/master/Tools/Verification/DuplicatiVerify.py

It performs the check externally, i.e. it does not know if the file is a valid zip file, but is is an extra validation. It works with and without encryption and is intended to be executed on the remote storage to guarantee integrity.

The dindex files are not required to restore the files (nor the local database), but they make everything significantly faster.

Would a similar approach work for what you (@balazer and @mnaiman) request?

I would think:
  
1. Backups have `--upload-verification-file` enabled
2. A new command, say `verify-integrity` is added
3. The new command downloads the `*-verification.json` file
4. It then checks all files are present, including size match verification
5. It then downloads all dlist and dindex files and verify their hashes
6. Optionally it can also do this with the dblock files

It can be added as an option to allow the backup to perform the check after each operation.

**Comment by balazer** [comment]
@kenkendk, ultimately what I am asking for is a way to know that I will be able to successfully restore my files, without running an actual restore and without downloading all of the data.  

What you've proposed is a check of the integrity of the remote storage, basically making sure that all of the files Duplicati thinks should be there are there.  The integrity of the remote storage is certainly important to the integrity of the backup.  I won't be able to restore all of my files if some of the backup volumes are missing or corrupt.  

But frankly I'm not too worried that files will mysteriously vanish from Google Drive.  I'm more worried about whether Duplicati will operate correctly, even in the face of factors outside of its control, like a loss of connectivity to the remote storage, a loss of connectivity to the source files, or the system being shut down or suspended right in the middle of a backup operation.  Even when the remote storage is operating perfectly, there's lots of other things that can go wrong.  Testing the integrity of the backup means going through all of the same actions that the system would to restore the files, except perhaps for actually downloading all of the dblock files.  Checking the sizes of the dblock files in the remote storage gives reasonable certainty that they contain the expected data.  A step further would be to retrieve the md5 hash of the dblock files, which is something that a lot of cloud storage platforms actually support. Integrity checking should also not rely on the local database, and instead try to recreate all of that data from the remove volumes. (since in a real restore, you might not have any local database)

Right now it seems that I just need to trust what Duplicati is doing, with no easy way to know that my backup can be relied upon.  The whole operation of Duplicati seems quite opaque.  

Contrast that with dclone.  While dclone of course is not a full-featured backup application, it can be used for backups in a limited way.  And very importantly with dclone, it's easy to be sure that the backup really reflects the source files.  For one thing, it doesn't use any encryption or repacking.  Every file in the backup is a one-to-one copy of a source file. I can list the backup files and see that they really match the source files.  After a backup completes, I can run dclone again with the same command, and it will check that all of the backup files really do match the source files, by checking the size and modification date, or even by checking the MD5 hash or SHA - and it can do that without downloading any backup files.  If any discrepancies are found, dclone will fix them automatically by re-uploading the mismatched files.  Thus a dclone backup always converges towards a clone of the source files.  The backup is self-repairing.  There's no database to be corrupted, and no consistency between the backup volumes to be maintained.  It's easy to verify the integrity of the backup, and easy to repair any problems that are found.

**Comment by kenkendk** [comment]
@balazer It depends on what you want, in terms of features. There are many nice tools out there, such as rclone, etc.

Duplicati works with a few assumptions that dictate how things are implemented:

1) Client side only; no server software installation, no server maintenance
2) Remote destination cannot be trusted; backups are encrypted and kept track of
3) Uploads/downloads are slow

I have added a guide that walks through the Duplicati backup process to make it less opaque:
https://www.duplicati.com/articles/Backup-Process/

If you store files as a copy, you are requesting a mirror/sync tool:
https://github.com/duplicati/duplicati/wiki/FAQ

If you want confidentiality for your files, you cannot simply store them as that would reveal filenames, sizes, directory structures, etc. It is also difficult to store different versions (latest version is broken, how do you get the one before?), both because of the naming issues, but also in terms of storage (each version is a new file).
Also, you will need a good connection to the remote storage, as each change in your files require a full file upload (there are corner cases where servers support patching files, but generally they do not).

rsync has a nice solution to the transfer/version issues, but it requires that you (or someone else) maintains a server with the rsync server software, and you cannot really get encryption on top.

In Duplicati you cannot guarantee that you can restore files, unless you restore files. I suggest that all users perform regular tests that they can restore their files to make sure.

To counter that problem, and reduce the risk of a surprising restore failure, Duplicati has a lot of checks and verifications that it performs on each run, such as checking if all remote files are present. It also does a sampling verification after each backup, where it downloads a few random files, and check that they are intact.

If you want to test the restore, you can perform a restore with the `--dry-run --no-local-db` flags set, to see that it actually works, without writing files to disk. I think we could add a `--no-block-download` flag that would also omit the downloads of the large `dblock` files, such that it happens much faster, but of course with less certainty for the restore.
